Fix `make ide-vscode` for Esp32/Windows

mikee47 opened the pull request:
- Fix 'missing g++' warning for Esp32 `make ide-vscode`
- Fix `tcp-serial-redirect` target for Windows
- Add Esp32 vscode GDB config
- Fix broken `gdb_do_break` for Esp32
- Add `util.mk` containing comming functions from `build.mk`, simplifies documentation Makefile
- Fix vscode task to ignore failed clean before build
- Add some #defined values to vscode browsing (e.g. COM_SPEED_SERIAL).
- Ignore headers for other architectures when browsing

mikee47 commented:
See #2407 for discussion of issues with vscode, Powershell and Windows.

Vity01 commented on Tools/ide/vscode/template/tasks.json:
```diff
@@ -24,14 +24,14 @@
             "detail": "Rebuild application and all Components",
             "type": "shell",
-            "command": "make clean components-clean SMING_ARCH=${command:cpptools.activeConfigName} && make -j ENABLE_GDB=",
+            "command": "make clean components-clean SMING_ARCH=${command:cpptools.activeConfigName}; make -j ENABLE_GDB=",
```
VSCode generator is used also for Linux, right? It must be distinguished for this OS. Probably using some variable.

mikee47 replied:
Good point. Maybe there's a cleaner way to list multiple commands...

Vity01 replied:
Looks like there can be an override for Windows specific:
https://code.visualstudio.com/docs/editor/tasks#_custom-tasks

Vity01 replied:
Btw. I just found the `;` works only for Poweshell :-). 
For Cmd.exe there should be && :-)). What a nice mess :-).

Vity01 replied:
I think the proper way might be a compound task:
https://code.visualstudio.com/docs/editor/tasks#_compound-tasks

mikee47 replied:
Compound tasks: exactly what I was looking for. Thank you!

Vity01 replied:
Excellent!

Vity01 commented on Tools/ide/vscode/template/tasks.json:
```diff
@@ -24,14 +24,14 @@
             "label": "Full rebuild (no debugging0",
```
typo (no debugging0 -> (no debugging)

Vity01 commented:
I checked the latest code and it still gives a warning:
```
c:\tools\Sming\samples\Basic_HwPWM>make ide-vscode

Basic_HwPWM: Invoking 'ide-vscode' for Esp8266 (debug) architecture
Using configuration 'standard'
C:\tools\Sming\Tools\ide\vscode/../common/
C:\tools\Sming\Tools\ide\vscode\template\intellisense/properties.json
C:\tools\Sming\Tools\ide\vscode\template\intellisense/configuration.json
Warning! '/C/tools/esp-quick-toolchain/gnu20/xtensa-lx106-elf/bin/xtensa-lx106-elf-g++.exe' doesn't exist
.vscode/tasks.json
C:\tools\Sming\Tools\ide\vscode\template\tasks.json
.vscode/launch.json
C:\tools\Sming\Tools\ide\vscode\template\launch.json
Warning! '/C/tools/esp-quick-toolchain/gnu20/xtensa-lx106-elf/bin/xtensa-lx106-elf-gdb.exe' doesn't exist
sming.code-workspace
C:\tools\Sming\Tools\ide\vscode\template\workspace.json
```

One more suggestion - the generated vscode profiles should be in order ESP, Host. Most likely you don't want to use a simulator as default.

mikee47 commented:
@Vity01 I can't quite see how this is happening. The only way the warning would show a path starting with `/C/...` instead of `C:/....` is if `check_path()` is failing on your system. Could you add some debug statements to `Tools/ide/sming.py` to see what's going on?

```
def find_tool(name):
    print(f"find_tool({name})")  # <<<<
    name = check_path(name)
    print(f"find_tool({name})")  # <<<<
    if sys.platform == 'win32':
        if os.path.splitext(name)[1] != '.exe':
            name += '.exe'
    print(f"find_tool({name})") # <<<<
...
```

What you *should* see is this:

```
...
find_tool(/C/tools/esp-quick-toolchain/gnu20/xtensa-lx106-elf/bin/xtensa-lx106-elf-g++)
find_tool(C:/tools/esp-quick-toolchain/gnu20/xtensa-lx106-elf/bin/xtensa-lx106-elf-g++)
find_tool(C:/tools/esp-quick-toolchain/gnu20/xtensa-lx106-elf/bin/xtensa-lx106-elf-g++.exe)
...
find_tool(/C/tools/esp-quick-toolchain/gnu20/xtensa-lx106-elf/bin/xtensa-lx106-elf-gdb)
find_tool(C:/tools/esp-quick-toolchain/gnu20/xtensa-lx106-elf/bin/xtensa-lx106-elf-gdb)
find_tool(C:/tools/esp-quick-toolchain/gnu20/xtensa-lx106-elf/bin/xtensa-lx106-elf-gdb.exe)
...
```

mikee47 commented:
@Vity01 I think your working directory isn't in sync - I'm seeing `C:\tools\Sming\Tools\ide\vscode/../common/` which was removed in e5879cbe837aa063586f53c82da7bb35ffeb4e6f

Vity01 commented:
Ahh, my bad, sorry. You are using (naturally) another branch for that. 
I will switch to another one and let you know.

Vity01 commented:
Excellent, the warnings are gone. 
I randomly selected sample `Basic_Tasks`, it's opened, includes are generated, but you get errors in IDE 
![image](https://user-images.githubusercontent.com/2512959/139813322-af1ac434-36ae-47a4-b309-4af24765349d.png)

Maybe project folder `include` should be also added to the include list if detected.
There are more projects with `include` folder.

mikee47 commented:
> Maybe project folder `include` should be also added to the include list if detected. (`${workspaceFolder}/include`)
> There are more projects with `include` folder.

I've added actual project directory recursively so will pick up most things.

Vity01 commented:
I confirm it's ok now.

slaff commented:
> I confirm it's ok now.

@mikee47 Is this PR ready for merging?

mikee47 commented:
@slaff Yep, all done.
